### PR TITLE
[Product Refactor] Move variant unit sizes to variant

### PR DESCRIPTION
**Pull request by rioug**
**⚠️ Please use clockify code https://github.com/openfoodfoundation/openfoodnetwork/issues/9105 REST API - Products when working on this ⚠️**

#### What? Why?

Last step of the first part of product refactor #9069
Move variant unit sizes to variant :
- variant_unit,
- variant_unit_scale,
- variant_unit_name

Review Note:

We should really refactor  app/webpacker/controllers/variant_controller.js and app/webpacker/controllers/variant_controller.js to extract anything that's common for the "unit popout". We could probably move it to component. This PR is already big enough so I did not want to do it here.

#### What should we test?

Anything that touches unit sizes (non exhaustive list), check that unit sizes are displayed properly and can be updated when applicable :
- Create a new Product
- Create a new Variant
- Update a variant
- Bulk product edit page legacy, create/update variant
- Bulk product edit page create/update variant
- Inventory page
- Producer shop page 
- order and fulfilment report 

Update variant page, I replaced angular with StimulusJS, beside checking unit sizes check there is no feature loss.

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [x] API changes (V0, V1, DFC or Webhook)
- [x] Technical changes only
- [ ] Feature toggled

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: app/models/product_import/entry_validator.rb [hunk 1]
```diff
@@ -312,6 +316,12 @@ def inventory_validation(entry)
     end
 
     def entry_matches_existing_variant?(entry, existing_variant)
+      # matching on the unscaled unit so we know if we are trying to update an existing variant
+      display_name_are_the_same?(entry, existing_variant) &&
+        existing_variant.unit_value == entry.unscaled_units.to_f
+    end
```
Comment (dacook) [inline app/models/product_import/entry_validator.rb]: I'm probably missing something but this seems unrelated to the refactor. Is this a bug fix?

File: spec/models/product_importer_spec.rb [hunk 2]
```diff
@@ -578,9 +578,12 @@
   describe "updating non-updatable fields on existing products" do
```
Comment (dacook) [inline spec/models/product_importer_spec.rb:578]: Worth testing non-updatable fields on variants now too?

File: app/models/product_import/entry_validator.rb [hunk 3]
```diff
@@ -305,7 +305,7 @@ def inventory_validation(entry)
         unscaled_units = entry.unscaled_units.to_f || 0
```
Comment (dacook) [inline app/models/product_import/entry_validator.rb:305]: Good one 👍

File: spec/services/sets/product_set_spec.rb [hunk 4]
```diff
@@ -69,22 +69,27 @@
...
+              }]
             }
           }
         end
 
         it 'updates the product without error' do
           expect(product_set.save).to eq true
-          expect(product_set.saved_count).to eq 1
+          # updating variant doesn't increment saved_count
+          # expect(product_set.saved_count).to eq 1
```
Comment (dacook) [inline spec/services/sets/product_set_spec.rb:90]: 🤔  Hmm, it probably should..

File: spec/javascripts/unit/admin/products/services/option_value_namer_spec.js.coffee [hunk 5]
```diff
@@ -1,29 +1,156 @@
 describe "OptionValueNamer", ->
```
Comment (dacook) [inline spec/javascripts/unit/admin/products/services/option_value_namer_spec.js.coffee:1]: Good find, thanks for consolidating 💯

File: app/assets/javascripts/admin/directives/maintain_unit_scale.js.coffee [hunk 6]
```diff
@@ -1,8 +0,0 @@
-angular.module("ofn.admin").directive "ofnMaintainUnitScale", ->
```
Comment (dacook) [inline app/assets/javascripts/admin/directives/maintain_unit_scale.js.coffee:1]: Yay, we don't need these files anymore it seems!

File: app/assets/javascripts/admin/bulk_product_update.js.coffee [hunk 7]
```diff
@@ -242,32 +241,28 @@ angular.module("ofn.admin").controller "AdminProductEditCtrl", ($scope, $timeout
     $window.location = destination
 
   $scope.packProduct = (product) ->
```
Comment (dacook) [inline app/assets/javascripts/admin/bulk_product_update.js.coffee:243]: "pack" means "prepare fields to save in the database", for example 
* "extracts variant_unit_with_scale into variant_unit and variant_unit_scale"
* "extracts unit_value and unit_description from unit_value_with_description"

https://github.com/openfoodfoundation/openfoodnetwork/pull/12787/commits/dc65c4a3595fbe59e851057d5893654051753e91#diff-c4c442b187d42bdc1fde8c83fa0c1211a6bb222914f3a0f82b9604fdb812bb13R524-R529 👇

**Review by dacook — COMMENTED** [review]
Reviewed first half(up to "Prettify javascript").

Sorry lots of comments, mostly just comprehension but I have a few queries as well.

**Review by mkllnk — COMMENTED** [review]
Review in progress:

* Fix DFC supplied product builder

File: app/models/spree/variant.rb [hunk 8]
```diff
@@ -71,21 +71,25 @@ class Variant < ApplicationRecord
...
                              if: proc { Spree::Config.products_require_tax_category }
 
+    validates :variant_unit, presence: true
     validates :unit_value, presence: true, if: ->(variant) {
-      %w(weight volume).include?(variant.product&.variant_unit)
+      %w(weight volume).include?(variant.variant_unit)
     }
-
     validates :unit_value, numericality: { greater_than: 0 }, allow_blank: true
-    validates :price, numericality: { greater_than_or_equal_to: 0 }
```
[the remark this answers is not in the document]
Comment (rioug) [reply app/models/spree/variant.rb:79]: duplicate code, see line 70

File: app/models/spree/variant.rb [hunk 9]
```diff
@@ -214,6 +221,23 @@ def total_on_hand
       Spree::Stock::Quantifier.new(self).total_on_hand
     end
 
+    # Format as per WeightsAndMeasures
+    # TODO test ?
+    def variant_unit_with_scale
+      scale_clean = ActiveSupport::NumberHelper.number_to_rounded(variant_unit_scale,
+                                                                  precision: nil,
+                                                                  strip_insignificant_zeros: true)
```
[the remark this answers is not in the document]
Comment (rioug) [reply app/models/spree/variant.rb]: It's addressed in a later commit

File: db/migrate/20240819045115_migrate_unit_size_to_variants.rb [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    SQL
+    )
+
+    ActiveRecord::Base.connection.execute(<<-SQL
+      UPDATE spree_variants
+      SET variant_unit_scale = spree_products.variant_unit_scale, variant_unit_name = spree_products.variant_unit_name
+      FROM spree_products
+      WHERE spree_variants.product_id = spree_products.id
+    SQL
+    )
```
[the remark this answers is not in the document]
Comment (rioug) [reply db/migrate/20240819045115_migrate_unit_size_to_variants.rb:19]: `variant_unit` already existed on both `product` and `variant` table, the condition is here to not override the value in the variant table. I don't think it's necessary for the other new fileds.

File: app/models/spree/product.rb [hunk 11]
```diff
@@ -45,35 +50,44 @@ class Product < ApplicationRecord
...
-    attr_accessor :price, :display_as, :unit_value, :unit_description, :tax_category_id,
-                  :shipping_category_id, :primary_taxon_id, :supplier_id
+    attr_accessor :price, :display_as, :unit_value, :unit_description, :variant_unit,
+                  :variant_unit_name, :variant_unit_scale, :tax_category_id, :shipping_category_id,
+                  :primary_taxon_id, :supplier_id
 
-    after_validation :validate_variant_attrs, on: :create
     after_create :ensure_standard_variant
-    after_update :touch_supplier, if: :saved_change_to_primary_taxon_id?
+    # after_update :touch_supplier, if: :saved_change_to_primary_taxon_id?
```
Comment (mkllnk) [inline app/models/spree/product.rb:89]: I guess that this needs removing?

File: app/models/spree/product.rb [hunk 12]
```diff
@@ -22,7 +22,12 @@ class Product < ApplicationRecord
     include LogDestroyPerformer
 
     self.belongs_to_required_by_default = false
-    self.ignored_columns += [:supplier_id, :variant_unit_scale, :variant_unit_name]
+    # These columns have been moved to variant. Currently this is only for documentation purposes,
+    # because they are declared as attr_accessor below, declaring them as ignored columns has no
+    # effect
```
Comment (mkllnk) [inline app/models/spree/product.rb:27]: > declaring them as ignored columns has no effect

I think that the important role of `ignored_columns` comes in handy when the columns will be deleted. If you haven't ignored them with a previous deploy then the app raises errors after the data migration removing the columns and before the app has been reloaded. This allows for zero-downtime deployment.

**Review by mkllnk — APPROVED** [review]
Epic! Great that you improved the code along the way and fixed little things.

I have to admit that I didn't read every line. But it looks pretty robust.

File: db/migrate/20240819045115_migrate_unit_size_to_variants.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (dacook) [reply db/migrate/20240819045115_migrate_unit_size_to_variants.rb:19]: Oh.. I must have missed that.. 
So if you can only select `variant_unit` on the product, and the `variant` is the source of truth, the code is already assuming these are in sync. So in theory, no variant_unit values will not need to be migrated (but it's still a good idea to have the migration just in case).

File: spec/system/admin/products_v3/update_spec.rb [hunk 13]
```diff
@@ -507,6 +511,30 @@
...
+          within new_variant_row do
+            expect_browser_validation('input[aria-label="Unit value"]',
+                                      "Please fill in this field.")
+          end
+
+          # Fix error
+          within new_variant_row do
+            fill_in "Unit value", with: "200"
+          end
+
```
Comment (dacook) [inline spec/system/admin/products_v3/update_spec.rb:538]: Nice specs!

File: app/models/product_import/entry_validator.rb (hunk 1 above)
Comment (rioug) [reply app/models/product_import/entry_validator.rb]: It looks like it, I can't really remember. I'll make sure it's in the notes to test product import

File: app/views/admin/shared/_tooltip.html.haml [hunk 14]
```diff
@@ -1,5 +1,8 @@
-%div{"data-controller": "tooltip"}
-  %a{"data-tooltip-target": "element", href: link, class: link_class}
+- tooltip_placement = defined?(placement) ? placement : "top"
+- tooltip_link = defined?(link) ? link : ""
+- tooltip_link_class = defined?(link_class) ? link_class : ""
+%div{"data-controller": "tooltip", "data-tooltip-placement-value": tooltip_placement }
+  %a{"data-tooltip-target": "element", href: tooltip_link, class: tooltip_link_class}
```
Comment (dacook) [inline app/views/admin/shared/_tooltip.html.haml:5]: I'm yet to find a prettier Ruby pattern for handling optional undefined variables in views..

File: app/views/spree/admin/variants/_form.html.haml [hunk 15]
```diff
@@ -98,7 +97,7 @@
       = f.collection_select(:primary_taxon_id, Spree::Taxon.order(:name), :id, :name, { include_blank: true }, { class: "select2 fullwidth" })
 
     .field
-      = f.label :supplier, t(:spree_admin_supplier)
+      = f.label :supplier, raw(t(:spree_admin_supplier) + content_tag(:span, ' *', :class => 'required'))
```
Comment (dacook) [inline app/views/spree/admin/variants/_form.html.haml:100]: I'm not sure, but the formbuilder might add the asterisk for us?
```suggestion
      = f.label :supplier, t(:spree_admin_supplier), required: true
```

But if not, perhaps a block will allow us to avoid the `raw` call:
```suggestion
      = f.label :supplier do
        = t(:spree_admin_supplier)
        = content_tag(:span, ' *', :class => 'required'))
```

File: spec/system/admin/variants_spec.rb [hunk 16]
```diff
@@ -13,18 +13,22 @@
...
+      click_on "Unit" # activate popout
+      # Unit popout
+      fill_in "Unit value", with: "1"
+      fill_in 'Price', with: 2.5
       select taxon.name, from: "variant_primary_taxon_id"
       select2_select product.variants.first.supplier.name, from: "variant_supplier_id"
+
       click_button 'Create'
 
       # Then the variant should have been created
```
Comment (dacook) [inline spec/system/admin/variants_spec.rb:34]: Hmm shouldn't we assert that the entered values were saved, too?

File: spec/system/admin/products_v3/update_spec.rb [hunk 17]
```diff
@@ -526,8 +526,10 @@
           # Client side validation
           click_button "Save changes"
           within new_variant_row do
+            # In CI we get "Please fill out this field." and locally we get
+            # "Please fill in this field."
```
Comment (dacook) [inline spec/system/admin/products_v3/update_spec.rb:529]: 😕  That's weird.
In the test environment, we use locale `en_TEST` which inherits everything from default local `en`.
In dev, we use `en_AU` (unless you have specified otherwise). I couldn't find any reason for this in `en_AU.yml`.

Hmm, but wait, when you run the tests locally it's still the same RAILS_ENV=TEST.

I give up, this fix looks good 👍

File: app/webpacker/controllers/edit_variant_controller.js [hunk 18]
```diff
@@ -160,13 +169,21 @@ export default class EditVariantController extends Controller {
...
-    let display = "block";
     if (this.variantUnit.value === "weight") {
-      display = "none";
+      return this.#hideWeight();
     }
 
+    // Show weight
     this.weight = this.element.querySelector('[id="variant_weight"]');
-    this.weight.parentElement.style.display = display;
+    this.weight.parentElement.style.display = "block";
```
Comment (dacook) [inline app/webpacker/controllers/edit_variant_controller.js:184]: Shouldn't this be conditional? it seems to overwrite whatever hideWeight did.

File: app/webpacker/css/admin_v3/pages/edit_variant.scss [hunk 19]
```diff
@@ -0,0 +1,55 @@
+@import "unit_popout";
+
+#edit_variant {
+
+  .popout {
+    @include unit_popout;
+
+    &__button {
+      // override popout button styles
```
Comment (dacook) [inline app/webpacker/css/admin_v3/pages/edit_variant.scss:9]: If we're refactoring, we shouldn't need to override or re-apply styles. 
Instead, can these be options to the unit_popout mixin? Or can the code that's related to the bulk products screen could be left on the bulk products css file?

File: app/models/spree/product.rb [hunk 20]
```diff
@@ -48,7 +48,27 @@ class Product < ApplicationRecord
     validate :validate_image
     validates :price, numericality: { greater_than_or_equal_to: 0, if: ->{ new_record? } }
 
-    accepts_nested_attributes_for :variants, allow_destroy: true
+    # These validators are used to make sure the standard variant created via
+    # `ensure_standard_variant` will be valid. The are only used when creating a new product
+    with_options on: :create_and_create_standard_variant do
```
Comment (dacook) [inline app/models/spree/product.rb:58]: Instead of creating a specific context name, can we just use Rails' `create` context which is automatically applied?

File: spec/system/admin/products_v3/update_spec.rb (hunk 17 above)
Comment (dacook) [reply spec/system/admin/products_v3/update_spec.rb:529]: 🤦 Ohhh, this is a browser message, not Rails. 
The browser message is dependent on the system locale. This will likely cause problems for devs in other countries, like Filipe and Ahmed. 

So I think we can't assert the message contents. But the JS API should at least provide a way to see if it's valid or not.

File: app/models/spree/product.rb (hunk 12 above)
Comment (dacook) [reply app/models/spree/product.rb:27]: Thanks Maikel, we should probably update/remove the comment then to avoid future misunderstanding.

**Review by dacook — COMMENTED** [review]
Phew, done! 
I have several comments, probably not all of them need changes, but I think some do (whether now or in a follow-up PR). 
I will review my own review and mark any required changes with a `±` in the comment.

File: spec/system/admin/products_v3/actions_spec.rb [hunk 21]
```diff
@@ -259,20 +259,6 @@ def expect_other_columns_visible
...
-
-          click_product_clone "Apples"
-
-          expect(page).to have_content "Unable to clone the product"
-
-          within "table.products" do
-            # Products does not include the cloned product.
-            expect(all_input_values).not_to match /COPY OF Apples/
-          end
-        end
```
[the remark this answers is not in the document]
Comment (dacook) [reply spec/system/admin/products_v3/actions_spec.rb]: `±` I think we need to keep this. Inevitably we will have invalid products in the system (eg what if validation rules change in the future).

File: spec/system/admin/products_v3/update_spec.rb (hunk 17 above)
Comment (dacook) [reply spec/system/admin/products_v3/update_spec.rb:529]: `±` I think we should fix this as it will cause confusion for new contributors.

File: spec/models/product_importer_spec.rb (hunk 2 above)
Comment (rioug) [reply spec/models/product_importer_spec.rb:578]: This is actually testing the non-updatable fields on the variant, anyway I revisited this and realised we could remove the check for non-updatable product fields.

File: spec/services/sets/product_set_spec.rb (hunk 4 above)
Comment (rioug) [reply spec/services/sets/product_set_spec.rb:90]: Indeed, this PR was not the place for it though

**Review by dacook — CHANGES_REQUESTED** [review]
I've added some `±` and one ❓ that I think require addressing, can you please check these? 
Other comments might be worth addressing too, but I don't consider them blocking.

File: app/models/spree/product.rb (hunk 12 above)
Comment (rioug) [reply app/models/spree/product.rb:27]: Indeed, I just wanted to highlight that currently it has no effect. Do you think I should update the comment ?

File: app/webpacker/controllers/edit_variant_controller.js [hunk 22]
```diff
@@ -0,0 +1,169 @@
...
+
+  //#showWeight() {
+  //  this.weight = this.element.querySelector('[id="variant_weight"]');
+  //  this.weight.parentElement.style.display= "block"
+  //}
+
+  //#hideWeight() {
+  //  this.weight = this.element.querySelector('[id="variant_weight"]');
+  //  this.weight.parentElement.style.display= "none"
+  //}
```
[the remark this answers is not in the document]
Comment (rioug) [reply app/webpacker/controllers/edit_variant_controller.js]: Taken care of in a later commit.

File: app/views/spree/admin/variants/_form.html.haml (hunk 15 above)
Comment (rioug) [reply app/views/spree/admin/variants/_form.html.haml:100]: I'll have a look, I think I copied that from somewhere else

File: app/models/spree/product.rb (hunk 12 above)
Comment (dacook) [reply app/models/spree/product.rb:27]: 🤷 
Probably not worth the time thinking about it. Marking as resolved ;)

File: spec/lib/spree/core/product_duplicator_spec.rb [hunk 23]
```diff
@@ -99,18 +98,6 @@
   end
 
   describe "errors" do
-    context "with invalid product" do
-      let(:product) {
-        # name is a required field
-        create(:product).tap{ |p| p.update_columns(variant_unit: nil) }
-      }
```
[the remark this answers is not in the document]
Comment (rioug) [reply spec/lib/spree/core/product_duplicator_spec.rb:106]: If you pass product with a nil name, it's not going to fail as the new product will have a name of "COPY OF " , and will be a valid product.

File: spec/system/admin/products_v3/actions_spec.rb (hunk 21 above)
[the remark this answers is not in the document]
Comment (rioug) [reply spec/system/admin/products_v3/actions_spec.rb]: I can't think of a way to get this to work .Even if you mock the product to return a `nil` name, the `ProductDuplicator` will set the name to "COPY OF " meaning it will create a valid cloned product. Any ideas ?

File: app/webpacker/controllers/edit_variant_controller.js (hunk 18 above)
Comment (rioug) [reply app/webpacker/controllers/edit_variant_controller.js:184]: If we are hiding this bit of code doesn't run, see the return in the code above:
```javascript 
    if (this.variantUnit.value === "weight") {
      return this.#hideWeight();
    }

```

File: app/models/spree/product.rb (hunk 20 above)
Comment (rioug) [reply app/models/spree/product.rb:58]: We could, but I would argue, in this scenario, it's better to be explicit with a specific context.

File: spec/lib/spree/core/product_duplicator_spec.rb (hunk 23 above)
[the remark this answers is not in the document]
Comment (dacook) [reply spec/lib/spree/core/product_duplicator_spec.rb:106]: 🤦  oh i see!
Another option then is to hit the maximum field length (256 chars).

[the remark this answers is not in the document]
Comment (rioug) [reply spec/lib/spree/core/product_duplicator_spec.rb]: Nice!  that does the trick ! Thanks for that.

File: spec/system/admin/products_v3/update_spec.rb (hunk 17 above)
Comment (rioug) [reply spec/system/admin/products_v3/update_spec.rb:525]: Fair enough, I ended removing the check for the message, I just check there is one present.

File: app/webpacker/css/admin_v3/pages/edit_variant.scss (hunk 19 above)
Comment (rioug) [reply app/webpacker/css/admin_v3/pages/edit_variant.scss:9]: From memory, it should be the other way around ie: the bulk products screen should be the one overriding the style because it's the one using a different colour. I was a bit pressed for time to go back and fix this. I'll raise a tech debt ticket when I have a minute, that can probably be a good first issue.

File: app/views/admin/shared/_tooltip.html.haml (hunk 14 above)
Comment (mkllnk) [reply app/views/admin/shared/_tooltip.html.haml:5]: I quite like to declare local variables at the top:

```haml
- link ||= ""
- link_class ||= ""
```

The alternative would be
```haml
- tooltip_link = local_assigns[:link].to_s
```

or

```haml
  %a{"data-tooltip-target": "element", href: local_assigns[:link], class: local_assigns[:link_class]}
```

File: spec/system/admin/products_v3/update_spec.rb [hunk 24]
```diff
@@ -706,4 +739,11 @@
       include_examples "updating image"
     end
   end
+
+  # Check a validation message is set, we don't check the message itself because the value is based
+  # on the browser's locale.
+  def expect_browser_validation(selector)
+    browser_message = page.find(selector)["validationMessage"]
+    expect(browser_message.present?).to be(true)
```
Comment (mkllnk) [inline spec/system/admin/products_v3/update_spec.rb:747]: I find the `be(true)` matcher confusing when it fails. I would either use `eq` or:
```suggestion
    expect(browser_message).to be_present
```

**Review by mkllnk — APPROVED** [review]

**Review by dacook — APPROVED** [review]
🏅 Awesome, an epic job!

File: spec/system/admin/products_v3/update_spec.rb (hunk 24 above)
Comment (rioug) [reply spec/system/admin/products_v3/update_spec.rb:747]: I always forget about these !

**Comment by drummer83** [comment]
@rioug There's a conflict here, sorry. Could you please have a look?
Also Github is asking for more reviews. Do we need that?

**Comment by rioug** [comment]
@drummer83 it's now ready to go.

**Review by dacook — APPROVED** [review]

**Comment by mkllnk** [comment]
@rioug, sorry, just created another merge conflict.

**Comment by rioug** [comment]
> @rioug, sorry, just created another merge conflict.

Sorted !

**Comment by rioug** [comment]
Simplecov failed, I think it's because it re ran only the failed jobs, maybe something we need to look into. 

Anyway, it doesn't block merging.

**Comment by drummer83** [comment]
Hi @rioug,

I'm sorry, I'm getting an error 500 on the /admin/products page on staging AU.
Bugsnag is here: https://app.bugsnag.com/yaycode/openfoodnetwork-aus/errors?filters[event.since]=30d&filters[error.status]=open&filters[app.release_stage]=staging

I'll move this back to In Progress until fixed. :pray:

**Comment by rioug** [comment]
@drummer83 There was some data and database related issue au `au_staging`, this is now fixed.

**Comment by rioug** [comment]
Hi @drummer83, see below

> * Variants created in current master have empty unit scale (and can't be cloned/saved). I created these variants directly before staging the PR to check if/how they would change after staging.

> * Older variants are ok. Variants created after staging the PR are ok. -> What's the difference to those from master?

From what I can tell it is a migration issue, causing a data issue. I was not able to reproduce this, but I think this it's because this PR has been staged before, so when you re staged it the data migration did not run, hence why you are seeing these issues. I reset the migration and deployed `v4.6.4` on `au-staging`. Could you try again ?

> * Variants created in current master are always displayed as mg/mL. Other variants automatically recalculate to g/kg/T or L/kL.

Again this is a data issue, there is a calculated field in the db that store "unit_presentation" for display purposes. The logic around unit value being displayed in the correct unit has not changed, I am not sure how you managed to have something like : "1000mg". I was not able to reproduce the issue, and looking at the data currently on `au-staging`, I couldn't find where it's coming from. Again it might be related to the migration not running as explained above. Could you check again ? If you manage to reproduce this, could you give details steps ?   

> * Bulk printing invoices doesn't work. Nothing happens at all, nothing in the console, nothing in Bugsnag.

I am assuming you are talking about the "Print invoices" from the action drop down on the "Orders" page ? I have not been able to reproduce this either, where you trying to print high number of invoices ? It's possible you hit something similar to the infinite loading bug we had on the report pages, but you should at least see a modal saying "Compiling Invoices"


> * Blank can be selected as unit scale -> should not be offered for selection.

This is to make sure a unit scale is selected by the user, but maybe it should be "Select Unit Scale" instead of a blank entry ?
@RachL what do you think ?

**Comment by RachL** [comment]
> but maybe it should be "Select Unit Scale" instead of a blank entry ?

@drummer83 @rioug I don't have time to stage it today, but unit scale is still mandatory, correct? Is the blank entry is only a problem because users might not see they need to add something before they hit save.

Given the size of the work here (and if my assumptions are correct) we could add a "select unit scale" info as part of a follow-up PR or something we can give as good-first-issue.

**Comment by filipefurtad0** [comment]
Hey @rioug ,

I've done some tests as a setup, before staging this PR.

image

image

For some reason, the shopfront displays the second variant as 0mL -> I'd expect 2mL as in the report and products page.

After staging, we're getting an error 500 on the products page:

https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/66ec7f100d87c81767cef62f?event_id=66ec7f1000f89217a3770000&i=sk&m=nw

Any idea why this is happening?
Moving back to in dev.

**Comment by rioug** [comment]
> After staging, we're getting an error 500 on the products page:
> 
> https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/66ec7f100d87c81767cef62f?event_id=66ec7f1000f89217a3770000&i=sk&m=nw

It usually a data issue when this error shows up, I 'll have a look

**Comment by rioug** [comment]
For some reason, the shopfront displays the second variant as 0mL -> I'd expect 2mL as in the report and products page.

That's an existing issue, I tracked it down to this code : 
https://github.com/openfoodfoundation/openfoodnetwork/blame/master/app/models/spree/line_item.rb#L223-L227

This is used to set `Spree::LineItem#unit_presentation` via `VariantUnits::OptionValueNamer#option_value_value_unit_scaled`, the problem happens because, in the case of a volume unit, the `final_weight_volume` is not set when the `LineItem` is not yet persisted so we end with a `LineItem#unit_value` of 0

**Comment by dacook** [comment]
It looks like the same thing mentioned here: https://github.com/openfoodfoundation/openfoodnetwork/pull/10812#issuecomment-1571992348
(Good to see that you are a consistent tester Filipe ;) ). I couldn't find any issue raised about this.

So I guess this means that the smallest digit for mL will always appear as `0`. Eg 15mL-> 10mL, 125mL -> 120mL, etc.
That's pretty confusing! Filipe would you be able to raise the issue?

**Comment by rioug** [comment]
@filipefurtad0 could you check this on `au-staging` or `fr-staging` ? I cleaned up the data on these two. I don't want to touch `uk-staging` as Maikel and Gareth are using it to test DFC order.

**Comment by filipefurtad0** [comment]
> It looks like the same thing mentioned here: https://github.com/openfoodfoundation/openfoodnetwork/pull/10812#issuecomment-1571992348

Hey @dacook . Thanks so much for reminding me of this issue!

> So I guess this means that the smallest digit for mL will always appear as 0. Eg 15mL-> 10mL, 125mL -> 120mL, etc.
That's pretty confusing! Filipe would you be able to raise the issue?

125mL is actually rounded to 130mL. I've opened an issue [here](https://github.com/openfoodfoundation/openfoodnetwork/issues/12874).

> could you check this on au-staging or fr-staging

@rioug sure, I'll re-test this issue on staging-AU then.

**Comment by filipefurtad0** [comment]
Hey @rioug ,

I've staged the PR on staging-FR now, the product page loads without any issues.

The issue with previously created variants having empty unit is fixed, which is great :+1: 

However, I can see some issues pointed by @drummer83 are still relevant, like being able to select an empty unit value, when creating an product in the products page:

image

So I thought the migration did not run properly. I've deleted the migration from the DB and restaged the PR in the hope to see this issue fixed, but [deployment failed](https://openfoodnetwork.slack.com/archives/CEF14NU3V/p1727475371847189), with the error:

https://app.bugsnag.com/open-food-france/coopcircuits/errors/66f72eaa626cc31c1047d655?event_id=66f72eaa00f87d88446d0000&i=sk&m=nw

I'm not sure how to proceed. Any idea on what the best approach is here? Added the feedback label.

**Comment by rioug** [comment]
Hi @filipefurtad0,
I am not sure how you "deleted the migration" but is look likes it wasn't done from a database perspective. I fixed the database on both `au-staging` and `fr-staging`. 

 FYI if you need to retest the migration, while this PR is staged, you can run the following to revert the migration done by the PR:
```
bundle exec rails db:migrate:down VERSION=20240819045115
bundle exec rails db:migrate:down VERSION=20240618013850
```


> However, I can see some issues pointed by @drummer83 are still relevant, like being able to select an empty unit value, when creating an product in the products page:

To be clear this has not been fixed so it's expected to be able to select an empty unit value (Rachel is happy for this to be fixed later on). However, if a variant was created before the PR, it shouldn't have an empty unit value after the PR has been staged.

**Comment by drummer83** [comment]
@rioug, conflicts again... :cry:

**Comment by filipefurtad0** [comment]
Hey @rioug,

I can confirm that this PR introduces the "possibility" to select an empty value from the list; However, it is not possible to save these changes, attempting to do so, triggers the warning below:

image

For comparison, before this PR:

image

This is valid for new and old variants, in the sense that an empty choice is appearing in the UI in both cases. However, all variants have attributed unit values.

> To be clear this has not been fixed so it's expected to be able to select an empty unit value (Rachel is happy for this to be fixed later on).

Alright, thank you for pointing this out. I'll open an issue on this. Thanks for all the effort and resilience. Merging (finally!) :tada:
